Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingBifurcatedImageBuffer.mm

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+#import "DynamicContentScalingBifurcatedImageBuffer.h"
+
+#if ENABLE(RE_DYNAMIC_CONTENT_SCALING)
+
+#import "DynamicContentScalingImageBufferBackend.h"
+#import <WebCore/DynamicContentScalingDisplayList.h>
+#import <WebCore/ImageBufferBackend.h>
+#import <wtf/MachSendRight.h>
+
+namespace WebKit {
+
+DynamicContentScalingBifurcatedImageBuffer::DynamicContentScalingBifurcatedImageBuffer(Parameters parameters, const WebCore::ImageBufferBackend::Info& backendInfo, const WebCore::ImageBufferCreationContext& creationContext, std::unique_ptr<WebCore::ImageBufferBackend>&& backend, WebCore::RenderingResourceIdentifier renderingResourceIdentifier)
+    : ImageBuffer(parameters, backendInfo, creationContext, WTFMove(backend), renderingResourceIdentifier)
+    , m_dynamicContentScalingBackend(DynamicContentScalingImageBufferBackend::create(ImageBuffer::backendParameters(parameters), creationContext))
```

Comment:
Can DynamicContentScalingImageBufferBackend::create() fail for any reason? I would suggest the following:

1. `RemoteRenderingBackend::createImageBuffer()` creates `DynamicContentScalingImageBufferBackend`. If the creation fails it will fall back to normal `ImageBuffer` or return an error.
2. `createImageBufferInternal()` will pass `dynamicContentScalingBackend` as the last argument to `ImageBuffer::create()`.
3. `ImageBuffer::create()` will pass it here to this constructor as the last argument.
4. This constructor will move it to `m_dynamicContentScalingBackend`

Reply:
Only for empty backendSize, and I'm not even sure why we do that.

I'll give this a shot.

Reply:
I don't love this, because it means passing dynamicContentScalingBackend even in the non-DCS case, or duplicating this code, or something? Going to defer this to later cleanup. Also I don't think it's critical because the main ImageBuffer backend will also fail to allocate for a superset of the reasons that the DCS one would (which is: only empty backendSize)